pkg/sdkserver/sdkserver.go

```diff
@@ -762,20 +790,202 @@ func (s *SDKServer) GetCounter(ctx context.Context, in *alpha.GetCounterRequest)
+		if in.CounterUpdateRequest.Capacity.GetValue() < 0 {
+			return nil, errors.Errorf("OUT_OF_RANGE. Capacity must be greater than or equal to 0. Found Capacity: %d", in.CounterUpdateRequest.Capacity.GetValue())
+		}
+		capacitySet := in.CounterUpdateRequest.Capacity.GetValue()
+		batchCounter.capacitySet = &capacitySet
+	default:
+		return nil, errors.Errorf("INVALID_ARGUMENT. Malformed CounterUpdateRequest: %v", in.CounterUpdateRequest)
+	}
+
+	s.gsCounterUpdates[name] = batchCounter
+
+	// Queue up the Update for later batch processing by updateCounters.
+	s.workerqueue.Enqueue(cache.ExplicitKey(updateCounters))
+	return &alpha.Counter{}, nil
+}
+
+// updateCounter updates the Counters in the GameServer's Status with the batched update requests.
+func (s *SDKServer) updateCounter(ctx context.Context) error {
+	if !runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
+		return errors.Errorf("%s not enabled", runtime.FeatureCountsAndLists)
+	}
+
+	gs, err := s.gameServer()
+	if err != nil {
+		return err
+	}
+	gsCopy := gs.DeepCopy()
+
+	s.logger.WithField("batchCounterUpdates", s.gsCounterUpdates).Debug("Batch updating Counter(s)")
+	s.gsUpdateMutex.Lock()
+	defer s.gsUpdateMutex.Unlock()
+
+	for name, ctrReq := range s.gsCounterUpdates {
+		counter, ok := gsCopy.Status.Counters[name]
+		if !ok {
+			continue
+		}
+		// If the Counter is the same as the original Counter stored in the counterUpdateRequest
+		// then no other changes have been made to this Counter, and we can overwrite the values.
+		if cmp.Equal(counter, ctrReq.counter) {
+			count := counter.Count
+			if ctrReq.countSet != nil {
+				count = *ctrReq.countSet
+			}
+			count += ctrReq.diff
+			counter.Count = count
+			if ctrReq.capacitySet != nil {
+				counter.Capacity = *ctrReq.capacitySet
+			}
+		} else {
+			// If the Counter is NOT the same as the original Counter stored in the counterUpdateRequest
+			// then changes have been made to the Counter since we validated the incoming changes in
+			// UpdateCounter, and we need to verify if the batched changes can be fully applied, partially
+			// applied, or cannot be applied.
+			if ctrReq.capacitySet != nil {
+				counter.Capacity = *ctrReq.capacitySet
+			}
+			if ctrReq.countSet != nil {
+				counter.Count = *ctrReq.countSet
+			}
+			newCnt := counter.Count + ctrReq.diff
+			if newCnt < 0 {
+				// If Count < 0 truncate requests by setting the Count to 0.
+				newCnt = 0
+				s.logger.Debug("truncating Count in Update Counter request to 0")
+			}
+			// TODO: This means there may be increment requests in the collapsed Diff that shouldn't have gone through. We may want a diffList for this case?
+			// If Count > Capacity and the Diff is decreasing the Count, reduce the Count by the Diff.
+			// Otherwise if Count > Capacity and the Diff is increasing the Count, truncate at Capacity.
+			if newCnt > counter.Capacity && ctrReq.diff > 0 {
```

Comment:
I think this can be simpler: If the count ends up below 0, then make it 0, if the count ends up above capacity, then make it capacity - same as the logic we have in `SetCounter` 👍🏻